### PR TITLE
docs: 3rd Party Library Design Doc.

### PR DESCRIPTION
**Pull request by hansl**

### CODE REVIEW

**Comment by jkuri** [comment]
thanks for this. lgtm

File: docs/design/third-party-libraries.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+The following use cases need to be supported by `ng install`:
+
+1. Support for _any_ thirdparties, ultimately falling back to running `npm install` only and
+doing nothing else, if necessary.
+1. Not a new package manager.
+1. Metadata storage by thirdparties in `package.json`. This must be accessible by plugins for
+the build process or even when scaffolding.
+1. Proper configuration of SystemJs in the browser.
+1. SystemJs configuration needs to be kept separated from the autogenerated part.
```
Comment (cironunes) [inline docs/design/third-party-libraries.md]: minor: SystemJ`S`

File: docs/design/third-party-libraries.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+1. The build process right now is faulty because the `tmp/` directory used by Broccoli is
+inside the project, which means that `node_modules/` is understood by Typescript. We need to
+proper map the imports and move them to `dist/`.
+1. Potentially hook into scaffolding.
+1. Potentially hook into the build.
+1. Safe uninstall path.
+
+# Usages
+
+Here's a few stories that should work right off:
```
Comment (filipesilva) [inline docs/design/third-party-libraries.md]: No angular2-material story?

File: docs/design/third-party-libraries.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+1. Support for _any_ thirdparties, ultimately falling back to running `npm install` only and
+doing nothing else, if necessary.
+1. Not a new package manager.
+1. Metadata storage by thirdparties in `package.json`. This must be accessible by plugins for
+the build process or even when scaffolding.
+1. Proper configuration of SystemJs in the browser.
+1. SystemJs configuration needs to be kept separated from the autogenerated part.
+1. The build process right now is faulty because the `tmp/` directory used by Broccoli is
+inside the project, which means that `node_modules/` is understood by Typescript. We need to
+proper map the imports and move them to `dist/`.
```
Comment (filipesilva) [inline docs/design/third-party-libraries.md]: I don't understand what you mean here at all, but might just be my fault. Can you dumb it down a bit for me?

File: docs/design/third-party-libraries.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+organization.** If so, run the steps above as if ng install angular/angular-${libName}. If
+this install fails, ignore the failure.
+
+  These packages can be used to wrap libraries that we want to support but can't update
+easily, like Jasmine or LESS.
+1. **Run `package["angular-cli"].scripts["postinstall"]` if it exists.**
+
+# Hooks
+The third party library can implement hooks into the scaffolding, and the build system. The
+CLI's tasks will look for the proper hooks prior to running and execute them.
```
Comment (filipesilva) [inline docs/design/third-party-libraries.md]: So prior to generation and build commands, the cli has to traverse the dependency tree? That sounds expensive, especially in watch mode.

Wouldn't it be better to just keep a reference to these hooks on install? Perhaps on `package["angular-cli"]["hooks"]`. That way order could be changed as well.

File: docs/design/third-party-libraries.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+with special values in the default string values. The values are taken from the `package.json`.
+
+We use a declarative style to enforce sensible defaults and make sure developers think about
+this. *In the case where developers want more complex interaction, they can use the
+install/uninstall hooks to prompt users.* But 3rd party libraries developers need to be aware
+that those hooks might not prompt users (no STDIN) and throw an error.
+
+# Providers
+Adding Angular providers to the app should be seamless. The build process will create a
+`providers.js` from all the providers contained in all the dependencies. The User can blacklist
```
Comment (filipesilva) [inline docs/design/third-party-libraries.md]: If the build process will generate this file, it will be hard for the user to do anything with it as his changes will be overwritten.

If it's intended to be a sort of configuration file, it would probably be better to keep it as a json file that is processed... except that I don't think we can do dynamic imports in TS (nor ES2015).

Manipulating this file will pose similar problems to the ones our current `ng install` has.

Comment (filipesilva) [inline docs/design/third-party-libraries.md]: Under this scheme, is there any way to add providers anywhere but at bootstrap? Or is that a manual task?

File: docs/design/third-party-libraries.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+providers it doesn't want.
+
+# Dependencies
+The install process should be recursive to allow for dependencies. However, the process should
+fail if there is a version difference between two libraries depending on the same library. This
```
Comment (filipesilva) [inline docs/design/third-party-libraries.md]: Shouldn't this sort of dependencies just be npm `peerDependencies`? Otherwise this sounds like recreating bower within angular-cli for angular 2 deps.

File: docs/design/third-party-libraries.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+to leave the liberty to the user to change the SystemJS configuration so that it fits their needs.
+
+We will not use SystemJS bundles in development. This is for better debugging, future proofing
+(when moving the build system) and better CDN support, as many of the loaded files will end up
+being pulled from a CDN in production. During the `ng build` process for production, the
+SystemJS configuration script will be rebuilt to fetch from the CDN.
+
+# Upgrade Strategy
+
+Upgrading older applications made using the CLI should be NOOP.
```
Comment (filipesilva) [inline docs/design/third-party-libraries.md]: What upgrade do you mean here?

File: docs/design/third-party-libraries.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+# Index.html
+
+We do not touch the `index.html` file during the installation task. The default page should
+link to a SystemJS configuration file that is auto generated by the CLI and not user
+configurable (see SystemJS below). If the user install a third party library, like jQuery, and
+wants to use it, they have to import it using `import * as $ from 'vendor/jquery'`.
+
+# SystemJS
+It is important that SystemJS works without any modifications by the user. It is also important
+to leave the liberty to the user to change the SystemJS configuration so that it fits their needs.
```
Comment (filipesilva) [inline docs/design/third-party-libraries.md]: How can the user change the SystemJS config if above you say it is not user configurable?

File: docs/design/third-party-libraries.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+```
+^(skip prompts for values passed on the command line)
+
+```sh
+$ ng install angularfire2 --quiet
+```
+^(using `--quiet` to skip all prompts and use default values)
+
+```sh
+$ ng install less
```
Comment (filipesilva) [inline docs/design/third-party-libraries.md]: This scenario sounds more like a CLI addon than a library to be used in an angular2 app. Not sure `ng install` should cater to it.

**Comment by filipesilva** [comment]
I really like the `appData` bit. Not so much the hooks bit though. 

I can't imagine that a lib that uses these hooks could be used without `angular-cli`, and that leads to a divide we are trying to avoid - having `angular2-*` versions of libs.

All these configuration options also seem to severely be overloading `package.json`. It is not hard to imagine that an app using 4-5 libraries would have the majority of it's `package.json` occupied by angular-cli module configuration. At that point, it's better to simply have an `angular-cli.json` file.

Perhaps the key issue here is the separation of cli addons vs simple libraries. Addons should change the functioning of the CLI, whereas simple libraries should be no more than imported and perhaps have some simple prompts for automatically add providers/configuration to an app.

File: docs/design/third-party-libraries.md (hunk 9 above)
Comment (jkuri) [inline docs/design/third-party-libraries.md]: With `ng install` we can install `less/sass/stylus` and enable the correlated CLI addon. I don't any see any repeating pattern on this three preprocessors so I think this will have to be hardcoded.

File: docs/design/third-party-libraries.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+wants to use it, they have to import it using `import * as $ from 'vendor/jquery'`.
+
+# SystemJS
+It is important that SystemJS works without any modifications by the user. It is also important
+to leave the liberty to the user to change the SystemJS configuration so that it fits their needs.
+
+We will not use SystemJS bundles in development. This is for better debugging, future proofing
+(when moving the build system) and better CDN support, as many of the loaded files will end up
+being pulled from a CDN in production. During the `ng build` process for production, the
+SystemJS configuration script will be rebuilt to fetch from the CDN.
```
Comment (filipesilva) [inline docs/design/third-party-libraries.md]: It feels a bit superfluous to have a strong commitment to SystemJS (for the CDN functionality) and still keep the brocolli build step. 

Perhaps we could move to SystemJS TS browser compilation for dev builds, and have the production build step simply compile TS to spare SystemJS from doing it.

File: docs/design/third-party-libraries.md (hunk 1 above)
Comment (hansl) [inline docs/design/third-party-libraries.md]: Done.

File: docs/design/third-party-libraries.md (hunk 3 above)
Comment (hansl) [inline docs/design/third-party-libraries.md]: Clarified, let me know.

File: docs/design/third-party-libraries.md (hunk 9 above)
Comment (hansl) [inline docs/design/third-party-libraries.md]: Build hooks can allow us to do that, though.

File: docs/design/third-party-libraries.md (hunk 4 above)
Comment (hansl) [inline docs/design/third-party-libraries.md]: Yes. The problem is if we copy and then `ng upgrade`, we need to rebuild a lot of information.

I don't think this will be affecting performance enough to be a problem. We can revisit.

File: docs/design/third-party-libraries.md (hunk 5 above)
Comment (hansl) [inline docs/design/third-party-libraries.md]: The user changes should be overwritten. Actually, the `providers.js` file will not be visible in the app; it's entirely code generated.

There are two sets of providers; the one the users add manually (between his components, for example), and one we auto generate at build time.

It's kept as information inside the `node_modules`.

File: docs/design/third-party-libraries.md (hunk 6 above)
Comment (hansl) [inline docs/design/third-party-libraries.md]: Yeah I'll remove this paragraph. Since we generate hooks and build depenedencies automatically, it can still work with `npm`.

File: docs/design/third-party-libraries.md (hunk 8 above)
Comment (hansl) [inline docs/design/third-party-libraries.md]: There's two SystemJS: the one in `index.html`, and another file that is non-configurable.

File: docs/design/third-party-libraries.md (hunk 10 above)
Comment (hansl) [inline docs/design/third-party-libraries.md]: We have no strategy for a production build, but it will involve concatenating + minifying, so it's safe to say we'll have to revisit broccoli.

File: docs/design/third-party-libraries.md (hunk 7 above)
Comment (hansl) [inline docs/design/third-party-libraries.md]: `ng install jQuery` -> New version of jQuery comes up -> `ng install` (or `ng upgrade`).

Comment (hansl) [inline docs/design/third-party-libraries.md]: Removed. This is more confusing.

**Comment by hansl** [comment]
@wesleycho you said you wanted to take a look at this proposal.
